File: hearth/ui/static/vad.js
/**
 * Client speech-band barge-in gate.
 * Thresholds must stay in sync with hearth/voice/vad.py.
 *
 * While the assistant is speaking, the outbound WebRTC mic track stays muted
 * until consecutive frames look like close speech (speech-band energy), not
 * broadband TV / HVAC / clatter. Outside playback the mic stays open.
 */
(function (global) {
  const SPEECH_HZ_LOW = 300;
  const SPEECH_HZ_HIGH = 3400;
  const MIN_RMS = 0.018;
  const MIN_SPEECH_BAND_RATIO = 0.42;
  const CONSECUTIVE_SPEECH_FRAMES = 4;
  const HANGOVER_FRAMES = 10;
  const POLL_MS = 40;

  function speechBandRatio(byteFreq, sampleRate, fftSize) {
    if (!byteFreq || !byteFreq.length || sampleRate <= 0 || fftSize <= 0) return 0;
    const hzPerBin = sampleRate / fftSize;
    let speech = 0;
    let total = 0;
    const limit = Math.min(byteFreq.length, Math.floor(fftSize / 2) + 1);
    for (let i = 0; i < limit; i += 1) {
      const mag = byteFreq[i] / 255;
      if (mag <= 0) continue;
      const energy = mag * mag;
      total += energy;
      const hz = i * hzPerBin;
      if (hz >= SPEECH_HZ_LOW && hz <= SPEECH_HZ_HIGH) speech += energy;
    }
    return total > 0 ? speech / total : 0;
  }

  function frameRms(timeData) {
    if (!timeData || !timeData.length) return 0;
    let sum = 0;
    for (let i = 0; i < timeData.length; i += 1) {
      const v = (timeData[i] - 128) / 128;
      sum += v * v;
    }
    return Math.sqrt(sum / timeData.length);
  }

  function isSpeechFrame(rms, bandRatio) {
    return rms >= MIN_RMS && bandRatio >= MIN_SPEECH_BAND_RATIO;
  }

  class BargeInGate {
    constructor(opts = {}) {
      this.consecutiveNeeded = opts.consecutiveNeeded || CONSECUTIVE_SPEECH_FRAMES;
      this.hangoverFrames = opts.hangoverFrames || HANGOVER_FRAMES;
      this.assistantSpeaking = false;
      this._speechStreak = 0;
      this._hangover = 0;
      this._open = true;
    }

    setAssistantSpeaking(speaking) {
      if (speaking === this.assistantSpeaking) return;
      this.assistantSpeaking = speaking;
      this._speechStreak = 0;
      if (speaking) {
        this._open = false;
        this._hangover = 0;
      } else {
        this._open = true;
        this._hangover = 0;
      }
    }

    observe(speechLike) {
      if (!this.assistantSpeaking) {
        this._open = true;
        this._speechStreak = 0;
        this._hangover = 0;
        return true;
      }
      if (speechLike) {
        this._speechStreak += 1;
        if (this._speechStreak >= this.consecutiveNeeded) {
          this._open = true;
          this._hangover = this.hangoverFrames;
        }
      } else {
        this._speechStreak = 0;
        if (this._open && this._hangover > 0) {
          this._hangover -= 1;
        } else if (this._open && this._hangover <= 0) {
          this._open = false;
        }
      }
      return this._open;
    }

    get micOpen() {
      return this._open;
    }
  }

  class SpeechBargeIn {
    /**
     * @param {MediaStreamTrack} track outbound mic track on the peer connection
     * @param {MediaStream} stream local mic stream (analysed; never played)
     */
    constructor(track, stream) {
      this.track = track;
      this.stream = stream;
      this.gate = new BargeInGate();
      this._ctx = null;
      this._analyser = null;
      this._freq = null;
      this._time = null;
      this._timer = null;
      this._running = false;
    }

    async start() {
      if (this._running) return;
      const Ctx = global.AudioContext || global.webkitAudioContext;
      if (!Ctx || !this.track || !this.stream) return;
      this._ctx = new Ctx();
      if (this._ctx.state === "suspended") {
        try {
          await this._ctx.resume();
        } catch (_) {
          /* ignore */
        }
      }
      const source = this._ctx.createMediaStreamSource(this.stream);
      this._analyser = this._ctx.createAnalyser();
      this._analyser.fftSize = 1024;
      this._analyser.smoothingTimeConstant = 0.35;
      source.connect(this._analyser);
      this._freq = new Uint8Array(this._analyser.frequencyBinCount);
      this._time = new Uint8Array(this._analyser.fftSize);
      this._running = true;
      this._applyMic(this.gate.micOpen);
      this._timer = global.setInterval(() => this._tick(), POLL_MS);
    }

    stop() {
      this._running = false;
      if (this._timer != null) {
        global.clearInterval(this._timer);
        this._timer = null;
      }
      if (this._ctx) {
        try {
          this._ctx.close();
        } catch (_) {
          /* ignore */
        }
        this._ctx = null;
      }
      this._analyser = null;
      this._applyMic(true);
    }

    setAssistantSpeaking(speaking) {
      this.gate.setAssistantSpeaking(Boolean(speaking));
      this._applyMic(this.gate.micOpen);
    }

    noteRealtimeEvent(type) {
      if (!type) return;
      if (
        type === "response.output_audio.delta" ||
        type === "response.audio.delta" ||
        type === "output_audio_buffer.started"
      ) {
        this.setAssistantSpeaking(true);
      }
      if (
        type === "response.done" ||
        type === "output_audio_buffer.stopped" ||
        type === "response.cancelled" ||
        type === "input_audio_buffer.speech_started"
      ) {
        // speech_started means barge-in already won — keep mic open for the turn.
        this.setAssistantSpeaking(false);
      }
    }

    _tick() {
      if (!this._running || !this._analyser) return;
      this._analyser.getByteFrequencyData(this._freq);
      this._analyser.getByteTimeDomainData(this._time);
      const rms = frameRms(this._time);
      const ratio = speechBandRatio(this._freq, this._ctx.sampleRate, this._analyser.fftSize);
      const speechLike = isSpeechFrame(rms, ratio);
      const open = this.gate.observe(speechLike);
      this._applyMic(open);
    }

    _applyMic(open) {
      if (this.track && this.track.enabled !== open) {
        this.track.enabled = open;
      }
    }
  }

  global.HearthVad = {
    SPEECH_HZ_LOW,
    SPEECH_HZ_HIGH,
    MIN_RMS,
    MIN_SPEECH_BAND_RATIO,
    CONSECUTIVE_SPEECH_FRAMES,
    HANGOVER_FRAMES,
    speechBandRatio,
    frameRms,
    isSpeechFrame,
    BargeInGate,
    SpeechBargeIn,
  };
})(typeof window !== "undefined" ? window : globalThis);

File: hearth/voice/vad.py
"""Speech-aware barge-in helpers.

OpenAI Realtime still owns turn detection, but Hearth raises the bar before
ambient energy (TV, HVAC, clatter) can interrupt playback:

1. Server session: ``noise_reduction`` filters the mic buffer *before* VAD.
2. Client (``/static/vad.js``): speech-band gate mutes the outbound WebRTC
   track while the assistant is speaking until consecutive speech-like frames
   are seen. Thresholds here are mirrored in that script — keep them in sync.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Sequence

# Mirrored in hearth/ui/static/vad.js
SPEECH_HZ_LOW = 300.0
SPEECH_HZ_HIGH = 3400.0
MIN_RMS = 0.018
MIN_SPEECH_BAND_RATIO = 0.42
CONSECUTIVE_SPEECH_FRAMES = 4
HANGOVER_FRAMES = 10
NOISE_REDUCTION_TYPE = "near_field"
SEMANTIC_EAGERNESS = "low"


def noise_reduction_config() -> dict[str, str]:
    """GA Realtime ``audio.input.noise_reduction`` — filters before VAD."""
    return {"type": NOISE_REDUCTION_TYPE}


def turn_detection_config() -> dict[str, Any]:
    """Semantic VAD tuned for less eager cut-ins; interrupt still on real speech."""
    return {
        "type": "semantic_vad",
        "eagerness": SEMANTIC_EAGERNESS,
        "create_response": True,
        "interrupt_response": True,
    }


def transcription_config() -> dict[str, str]:
    """GA Realtime input transcription for live user turns in the phone UI."""
    return {"model": "gpt-4o-mini-transcribe"}


def audio_input_config() -> dict[str, Any]:
    return {
        "noise_reduction": noise_reduction_config(),
        "transcription": transcription_config(),
        "turn_detection": turn_detection_config(),
    }


def speech_band_ratio(
    magnitudes: Sequence[float],
    *,
    sample_rate: float,
    n_fft: int,
) -> float:
    """Fraction of spectral energy in the speech band (300–3400 Hz).

    ``magnitudes`` are non-negative linear magnitudes for bins ``0 .. n_fft/2``.
    Broadband / low-frequency noise scores low; voice-shaped spectra score high.
    """
    if not magnitudes or sample_rate <= 0 or n_fft <= 0:
        return 0.0
    hz_per_bin = sample_rate / float(n_fft)
    speech = 0.0
    total = 0.0
    limit = min(len(magnitudes), n_fft // 2 + 1)
    for i in range(limit):
        mag = float(magnitudes[i])
        if mag <= 0:
            continue
        energy = mag * mag
        total += energy
        hz = i * hz_per_bin
        if SPEECH_HZ_LOW <= hz <= SPEECH_HZ_HIGH:
            speech += energy
    if total <= 0:
        return 0.0
    return speech / total


def is_speech_frame(
    rms: float,
    band_ratio: float,
    *,
    min_rms: float = MIN_RMS,
    min_ratio: float = MIN_SPEECH_BAND_RATIO,
) -> bool:
    """True when a single analysis frame looks like close speech, not ambient noise."""
    return rms >= min_rms and band_ratio >= min_ratio


@dataclass
class BargeInGate:
    """Mic open/closed while the assistant is speaking.

    Outside assistant playback the gate stays open (normal turn-taking).
    During playback the outbound track stays muted until consecutive speech-like
    frames arrive, then stays open through a short hangover.
    """

    consecutive_needed: int = CONSECUTIVE_SPEECH_FRAMES
    hangover_frames: int = HANGOVER_FRAMES
    assistant_speaking: bool = False
    _speech_streak: int = 0
    _hangover: int = 0
    _open: bool = True

    def set_assistant_speaking(self, speaking: bool) -> None:
        if speaking == self.assistant_speaking:
            return
        self.assistant_speaking = speaking
        self._speech_streak = 0
        if speaking:
            self._open = False
            self._hangover = 0
        else:
            self._open = True
            self._hangover = 0

    def observe(self, speech_like: bool) -> bool:
        """Feed one frame; returns whether the outbound mic should be enabled."""
        if not self.assistant_speaking:
            self._open = True
            self._speech_streak = 0
            self._hangover = 0
            return True

        if speech_like:
            self._speech_streak += 1
            if self._speech_streak >= self.consecutive_needed:
                self._open = True
                self._hangover = self.hangover_frames
        else:
            self._speech_streak = 0
            if self._open and self._hangover > 0:
                self._hangover -= 1
            elif self._open and self._hangover <= 0:
                self._open = False

        return self._open

    @property
    def mic_open(self) -> bool:
        return self._open
